app/assets/scripts/components/documents/pdf-preview/index.js

```diff
@@ -69,28 +47,76 @@ class AfterRenderHandler extends Handler {
     }
 
     const tocElement = content.querySelector('#table-of-contents');
-    const sectionHeadings = content.querySelectorAll('h2');
-
-    Array.from(sectionHeadings).forEach((h) => {
-      const section = document.createElement('div');
-      section.classList.add('toc-section');
-      tocElement.append(section);
-
-      const sectionTitle = getTitleElement(h.id, h.innerText);
-      section.append(sectionTitle);
-
-      const parent = h.parentNode;
-      const subHeadings = parent.querySelectorAll('h3');
-      if (subHeadings.length > 0) {
-        const subHeadingsContainer = document.createElement('div');
-        subHeadingsContainer.classList.add('sub-headings-container');
-        section.append(subHeadingsContainer);
-        Array.from(subHeadings).forEach((sh) => {
-          const subHeading = getTitleElement(sh.id, sh.innerText);
-          subHeadingsContainer.append(subHeading);
+
+    function generateSubHeadings(
+      currentHeading,
+      currentLevel,
+      sectionContainer,
+      parentContainer,
+      parentHeadingNumber
+    ) {
+      const currentSubHeadings = [];
+      let currentEl = currentHeading;
+
+      while (
+        currentEl.nextElementSibling !== null &&
+        currentEl.nextElementSibling.tagName !== `H${currentLevel}`
+      ) {
+        if (currentEl.nextElementSibling.tagName === `H${currentLevel + 1}`) {
+          currentSubHeadings.push(currentEl.nextElementSibling);
+        }
+
+        currentEl = currentEl.nextElementSibling;
+      }
+
+      if (currentSubHeadings.length > 0) {
+        const subHeadingSections = document.createElement('div');
+        subHeadingSections.classList.add('toc-section');
+        sectionContainer.append(subHeadingSections);
+
+        currentSubHeadings.forEach((subHeading, i) => {
+          const headingNumber = `${parentHeadingNumber}${i + 1}.`;
+          if (subHeading.children[0]) {
+            subHeading.children[0].prepend(`${headingNumber} `);
+          } else {
+            subHeading.prepend(`${headingNumber} `);
+          }
+          const subHeadingTitle = getTitleElement(
+            subHeading.id,
+            subHeading.innerText
+          );
+          subHeadingSections.append(subHeadingTitle);
+
+          generateSubHeadings(
+            subHeading,
+            currentLevel + 1,
+            subHeadingSections,
+            subHeading.parentNode,
+            headingNumber
+          );
         });
       }
-    });
+    }
+
+    function generateHeading() {
+      const sectionHeadings = content.querySelectorAll('h2');
+      Array.from(sectionHeadings).forEach((h, i) => {
```

Comment:
Minor: can we call the variable `heading` or something, instead of `h` ?

Reply:
^ That's actually not a strong opinion, since `h` in this context is `h1`, `h2`, so this is fine.